File: data_structures/linked_list/return_last_k.cpp
#include <stack>
#include "linked_list.cpp"

using namespace std;

template <class T> Node<T> *return_last(MyLinkedList<T> *list, int k) {
    if (k <= 0) {
        return NULL;
    }
    stack<Node<T>*> s;
    Node<T> *current = list->front();
    while (current != NULL) {
        s.push(current);
        current = current->next;
    }
    while (!s.empty()) {
        k--;
        if (k == 0) {
            return s.top();
        }
        s.pop();
    }
    return NULL;
}

template <class T> void print_last_k(MyLinkedList<T> *list, int k) {
    Node<T> *element = return_last(list, k);
    cout << "Looking for last " << k << "th" << endl;
    if (element == NULL) {
        cout << "Invalid index" << endl;
    } else {
        cout << "Element found '" << element->item << "'" << endl;
    }
}

int main() {
    MyLinkedList<int> list = MyLinkedList<int>();
    list.insert(6);
    list.insert(5);
    list.insert(4);
    list.insert(3);
    list.insert(2);
    list.insert(1);
    list.print_list();
    print_last_k(&list, 1);
    print_last_k(&list, 0);
    print_last_k(&list, 6);
    print_last_k(&list, 7);
    print_last_k(&list, 100);
    print_last_k(&list, -1);
    print_last_k(&list, 2);
    return 0;
}

File: ctci/chapter2/return_last_k.py
#!/bin/python

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


def last_k(head, k):
    current = head
    n = 0
    while current != None:
        n += 1
        current = current.next
    current = head
    l = n - k
    if l < 0:
        return None
    while l:
        current = current.next
        l -= 1
    return current


def create_linked_list(A):
    if A == []:
        return None
    x = A.pop(0)
    head = Node(x)
    current = head
    while A:
        x = A.pop(0)
        current.next = Node(x)
        current = current.next
    return head


def print_list(head):
    current = head
    while current != None:
        print "{0}".format(current.value)
        current = current.next


def test(head, k):
    print "=================================="
    print "Before"
    print_list(head)
    print "Last '{1}': '{0}'".format(last_k(head, k).value, k)


head = create_linked_list([1, 2, 3, 4, 5, 6])
test(head, 1)
head = create_linked_list([1, 2, 3, 4, 5, 6])
test(head, 3)
head = create_linked_list([1, 2, 3, 4, 5, 6])
test(head, 6)
